pageserver/src/page_cache.rs

```diff
@@ -486,12 +509,17 @@ impl PageCache {
         let mut shared = self.shared.lock().unwrap();
 
         // Can't move backwards.
-        assert!(lsn >= shared.last_valid_lsn);
+        let oldlsn = shared.last_valid_lsn;
+        if lsn >= oldlsn {
 
-        shared.last_valid_lsn = lsn;
-        self.valid_lsn_condvar.notify_all();
+            shared.last_valid_lsn = lsn;
+            self.valid_lsn_condvar.notify_all();
 
-        self.last_valid_lsn.store(lsn, Ordering::Relaxed);
+            self.last_valid_lsn.store(lsn, Ordering::Relaxed);
+        } else {
+            warn!("attempted to move last valid LSN backwards (was {:X}/{:X}, new {:X}/{:X})",
+                  oldlsn >> 32, oldlsn & 0xffffffff, lsn >> 32, lsn & 0xffffffff);
```

Comment:
Looks like we need an Lsn type that implements `Display`?

Reply:
yes! I could swear we had a github issue to track that already, but can't find one now. Created one now: https://github.com/zenithdb/zenith/issues/49